### ISSUE TITLE
Altimeters differences & setting

### ISSUE THREAD

**Issue by pturri** [fhorinek/SkyDrop#27]
1) how is absolute altidute calculated at start up? Based on GPS data only? Or pressure+GPS?
2) can you briefly explain the difference between Alt's?
Alt1 = Absolute to QNH1
Alt2 = Relative to Alt1
Alt3 = Relative to Alt1
Alt4 = Absolute to QNH2

**Comment by fhorinek** [comment]
Currently you cannot use GPS to calibrate altitude. QNH1 and QNH2 have to be set by user.
* Alt1 is calculated from pressure and QNH1 (actual pressure reduced to sea level for your actual location)
* Alt2 is Alt1 + delta. So after you set correct QNH1 you can use it for example as height above landing place. For example you are in 1700m and you know that landing is in 1200m. So you set delta to -500. Or just zero it on landing place using long click when you have the widget selected. 
* Alt3 is the same as Alt2 only with different delta (for alternative landing or something else)
* Alt4 is using QNH2, you can set any QNH here but lot of pilot will use 1013.25hPa, because the flight levels are defined at this level.

**Comment by pturri** [comment]
Ok thanks, this was what I supposed but better to get your confirmation :) 

Last question: are you planning to implement GPS autocalibration on next updates?

**Comment by fhorinek** [comment]
This might be more complicated. The GPS altitude is not the very precise. I would rather implement manual calibration using GPS, so the user will decide if he want to use it or not.

**Comment by pturri** [comment]
You are right, and I agree, IMHO anyway being GPS already on board it's a pity to do not exploit this information, there are some quite important advantages by using "also" GPS altitude:

- you may happen to takeoff from unusual places (H&F pilots on top) which height is not necessarily known. Gps altitude estimation might be wrong, but it will be anyhow much precise than any rough guess.

- always talking about H&F, let's say you set baro altimeter at your start. Then you walk up the hill for 3-4 hours. It is quite sure that atmospheric pressure is changed in the meanwhile. 

- Let's suppose you set the baro at takeoff, you fly for a while, when atmospheric pressure drops due to approaching bad weather. Any barometric altimeter will be severely affected by this pressure change, shown altitude might get very wrong and (even worse) real altitude might be way lower than shown one. In such situations, GPS error, despite not negligible, will be anyway better than a super precise pressure altitude, but calibrated 4 hours prior.

- moreover, by comparing pressure and GPS altitude you can acknowledge in a flash a severe atmospheric pressure change: when the two values get very different, it is clear that something is going on. Some outdoor GPS feature an alert when this happens.

As said, being GPS height already available, I'd suggest to add a further altimeter field showing GPS height, then you can decide to which Alt to rely more on depending on daily needs.

**Comment by fhorinek** [comment]
I agree, and I also like the baro alert function

**Comment by fhorinek** [comment]
TODO:
Set Alt from GPS - Easy

What about the alert function?
Say when Alt1 changes by 300m (configurable), then Alert pilot and set Gps alt to Alt1?
What do you think?

**Comment by pturri** [comment]
I'm glad you accepted my small idea :) 

Regarding alert, I think there are more approaches, depending on what you want to bring to pilot attention:

1) Takeoff altitude wrong: 
GPS <-> Alt1 check at boot -> if 2 values too different  = Alarm -> confirm to Set Gps alt to Alt1

2) pressure change warning: 
GPS <-> Alt1 timed check (30min?) -> if gap gets too big during the sampling period = Alarm "Severe pressure change!" -> confirm to Set Gps alt to Alt1 (and land... :) 
Being this ALSO a weather alert, you should find (or calculate) which is a reasonable pressure change per time unit that might lead to a weather degradation. I don't know this data, but it should not be hard to find. I can ask to a friend, he is a weather guru :)
Thes idea is this: 
let's say your take off is at 1000m, in Standard conditions baro should read 878hPa.
You switch Drop on e set GPS => Alt1 
After 30 min, you are still waiting to take off, GPS reads always 1.000m, but Alt1 = 1300m
Or, alike, you take off immediately, after 30 min GPS reads 2.000m (=760hPa) but Alt1 shows 2300m (729hPa)
Besides the altitude mistake (not negligibe), that you can easily fix in flight, pressure dropped more than 30hPa in 30 min. This is definetly worrisome.... 
The only issue is how to get a GPS data to make warning reliable.A spot reading might be very wrong, 300m meteres are possibile, but if you average a certain sample I'm quite sure you can get a very precised GPS data. 
Note: actually this will make Drop the first vario with such weather warning ;) 

3) In any case, I'd suggest to add an additional Alt5 (or AltG :) for GPS altitude value, at pilots taste.

**Comment by fhorinek** [comment]
So I will implement it like this
1. When GPS got 3D fix (or after ~50 3D fixes), the alert will pops out when difference will be more then 100m (configurable), 

2. I guest the simplest and most reliable method will be configurable altitude difference. 

3. Done already

**Comment by pturri** [comment]
I think you are doing it the right way :+1:

**Comment by SonnenSimon** [comment]
for the pressure change warning:
a quick weather change leads to a pressure change of 1-2 hpa/h. More than 2hpa/h means really stormy weather coming. This means an "altitude change" of 8-16m per hour. (1hpa -> about 8m)
So a big average of the gps height is neccesary to avoid false alarms.

Thanks for implementing all those amazing features!

**Comment by pturri** [comment]
thanks for the update, you are really making this device wonderful!


2015-09-08 11:07 GMT+02:00 SonnenSimon <notifications@github.com>:

**Comment by moyesboy** [comment]
Altitude setting from GPS is working really well on 6030.
It seems the altitude coming from GPS is averaged in the GPS chip, sop on 6030 (if you set for altimeter to be set by gps, whi9ch is default) the altitude flashes until you launch and then setting is fixed. So you can see GPS altitude, compare to map/notice on launch etc. It is very repeatable and accurate in my experience (but I am starting instrument maybe 15 mins before flight every time, because I am a hang glider and I take longer to prepare!).
I find setting like this gives me very good accuracy for altitude of even distant landing places after long flights.

Regarding pressure changes during the flight - I don't think even hang gliders can fly in weather where they cross a significant number of isobars on the pressure chart without being in danger!

This is much more useful than setting from a map (either printed or a terrain map in the instrument or another instrument) because normally we are on a steep slope and map altitude is hopeless - unless you are on an flat airfield.

**Comment by Fulminao** [comment]
Hi,
I've a question related to GNH2 altimeter. I don't know why, probably testing, the various options, I've changed the QNH2 value to 1013.19. I've tried to change this value to obtain the original value (1013.25) but I failed to do so, the only value that I can put in are 1013.19 or 1013.44 .
There is a way to put in the original value 1013.25?
Thanks.

**Comment by maticus** [comment]
I fully agree with moyesboy. The feature (for relative altimeters) like "set to GPS upon fix" would be great. Of course not after first 3D fix, but after a few (a minute or so).

**Comment by bubeck** [comment]
An additional Option would be, to set ALT1 based upong GPS Position and AGL (gound Level data stored on SD Card). Surely, this is only possible, if AGL data is available.

After getting a (stable) GPS fix, we could offer the user the following question?
  ALT1 is 1040m
 GPS height Shows 1208m
  AGL is 1180m

```
Set ALT1 to 
GPS   Nothing  AGL?
<        O       >

```

**Comment by maticus** [comment]
IMHO a nice option, but not so accurate on steep slopes and the elevation map needs to be good, as the error is approximately equal to: (2D fix err + elevmap position error) * tan(slope angle) + elevmap height error.

Not so obvious which is more accurate (GPS height vs GPS 2d + elevation map), but if we could choose among setting to "GPS height" and "GPS 2d + elevmap" it would be great.

Anyway, PLEASE make an option of automatic setting of altimeters upon (stable) fix. I like to configure my instrument at home, and then at the site I just turn it on, give it some time when I'm preping for launch and fly. Without having to touch anything on the vario :-)

**Comment by bluearcus** [comment]
An 'set baro altimeter from GPS on first fix' feature would be my preferred option. Average the GPS height over a short period after first fix, then set. Much better than DEM setting. As maticus suggests... fire and forget setting on the Altimeter config itself, not something that demands interaction with the instrument before takeoff.

**Comment by bubeck** [comment]
Feature is implemented and merged. Probably part of the next release. The functionality is as follows:

SkyDrop has up to 3 methods to obtain the correct height:
1. Value from barometer (possible error is, that air pressure naturally changes over time)
2. Value from GPS 3d fix (also has an error depending on quality of signal and receiver)
3. [optionally] Value from DEM at GPS 2d position (error is the quality of the data, the resolution and how steep the ground is)

The implementation associated an error value with each source and uses the value from the source with the least error. This typically means:

1. If you are not flying and GPS 2d fix with DEM and area is "flat": Use this as a starting point with a small error.
2. During flight time, keep using the barometer value for typically 3 hours. After that GPS 3d is more precise and will be used for re-calibration during flight

Feature must be turned on in Settings>Altimeters>Alt1 (Auto calibrate: YES).

**Comment by bluearcus** [comment]
How does the implementation determine relative errors of 2. and 1?

Most flying will occur in areas that are not flat, indeed, the vario will normally be switched on at launch, so by definition the DEM approach will generally be discarded immediately.

GPS 3d has a certain reasonable (95%) accuracy expectation, depending upon VDOP figure. Averaging over successive positions can improve this. For a good VDOP value (1 or less), errors should normally be less than 75ft.

Barometer though, no idea. If last flying was on a standard atmospheric day of 1013.2hPa but today is on a high pressure day at 1023hPa, then a 300ft under reading error would be expected. How can the algorithm make any judgement about the validity of a previous flying day's pressure setting in comparison to today?

I would suggest that unless VDOP is quite high (above 2.5) an average GPS height should be the approach. Where VDOPis above this, and no surface calibration was possible then stick to the calibrated baro, but correct this as soon as VDOP becomes good enough...

**Comment by bubeck** [comment]
Following methods are used for the error of
1. If SkyDrop is turned on, then barometer gets a (default) associated error of 100m. This would mean, that all other sources are normally better and will be used, whenever available. When the barometer is calibrated, the error from the better source is transfered to vario as its new associated error. Barometer error increases over time so that after 3 hours, it has an error of 100m.
2. Indeed VDOP is used, however, the used GPS receiver never sends something better than 2.3.

Hope that answers your questions. Currently, the implementation works quite well. I suggest we try it and improve it over time. Many suggestions are welcome.

**Comment by moyesboy** [comment]
Only std pressure is recorded in the baro field of the igc. So no calibrations affect the igc recording.In theory you might fly a long way to a remote airfield, call them on radio for a pressure altitude number, and then reset your altimeter so it reads correctly for your landing approach. :)Well, you might, if you were a sailplane.On 2 Jul 2018 10:26 am, bluearcus <notifications@github.com> wrote:OK, sow how I guess this will work in flight scenario...
Takeoff is on a hill... so DEM model is discarded.
Barometer has 100m error estimate and GPS signal is OK, so
GPS height is used to calibrate baro and
Error is (perhaps) 30m which is transferred to Baro.
Baro error now grows as flight proceeds.
On a regular basis Baro error will be found to be greater than GPS error and a recalibration will occur.
Will this not lead to frequent height discontinuities in the log files? And vario 'glitches'?
I'd prefer the initial calibration to be left alone once it has been made, and also after takeoff have no recalibrations if the Skydrop is making an IGC log.

**Comment by bubeck** [comment]
You got it. :)

Two ideas:
* it would be better to turn on SkyDrop some time before the start (e.g. when you walk to the starting point). This would give it a chance to recalibrate based on DEM whenever you cross a "flat" area. Apart from that, you are right
* Yes, it will recalibrate during flight with GPS 3d (every few hours). We can enter a comment into IGC to explain this. I think that a recalibration is better, than a wrong altitude... Maybe we find some better handling of glitches.

You should try it out and report, what you think should be done to improve.

**Comment by moyesboy** [comment]
Igc is recording GPS altitude and standard pressure altitude, so its numbers are not affected by recalibration, so recording the recalibration is not recording anything that is actually affecting the igc recording, unless you added extra user set altitude fields that are not part of the civil specification.I don't understand and worry that you are recalibrating pressure altimeter that pilot is referring to using GPS altitude because while flying gps altitude is not accurate at all. During climb it climbs too fast, recording a considerably higher max altitude than barometric. If you drop low from launch and climb back up it records much lower than you went, usually under the ground. This is the case for all the GPS devices I have tested and is clearly visible from just about all igc  files you might see from gliding flight.So GPS altitude while flying, unless you are cruising level with a motor, is not good for the pilot to refer to.If you recal  altimeter with gps at top of last climb you will arrive too low at goal every time.Changes in pressure are usually small, because we are not capable of flying in weather where we can fly across many isobars on the weather map. Very long distances are along the isobars anyway ;)So using a fixed barometric altitude from launch is usually best. Base map if it is flat, GPS averaged over time if it is sloping.On 2 Jul 2018 10:45 am, "Dr. Tilmann Bubeck" <notifications@github.com> wrote:You got it. :)
Two ideas:
it would be better to turn on SkyDrop some time before the start (e.g. when you walk to the starting point). This would give it a chance to recalibrate based on DEM whenever you cross a "flat" area. Apart from that, you are rightYes, it will recalibrate during flight with GPS 3d (every few hours). We can enter a comment into IGC to explain this. I think that a recalibration is better, than a wrong altitude... Maybe we find some better handling of glitches.
You should try it out and report, what you think should be done to improve.

**Comment by bluearcus** [comment]
As the Geoid GPS height and baro reading are in the IGC, then the IGC should be fine. However altering the QNH setting for a barometric altimeter in flight automatically is not normal behaviour and I would rather the instrument can be configured to calibrate the altimeter QNH to GPS or DEM altitude, if in a flat area and on the ground, when it has a good signal and to leave it set to this value for the period of the flight.Adding altimeter labels to indicate type (GPS, auto-baro, baro, QFE) would be useful. As would widgets for VDOP and HDOP.On 2 Jul 2018 14:16, moyesboy <notifications@github.com> wrote:Igc is recording GPS altitude and standard pressure altitude, so its numbers are not affected by recalibration, so recording the recalibration is not recording anything that is actually affecting the igc recording, unless you added extra user set altitude fields that are not part of the civil specification.I don't understand and worry that you are recalibrating pressure altimeter that pilot is referring to using GPS altitude because while flying gps altitude is not accurate at all. During climb it climbs too fast, recording a considerably higher max altitude than barometric. If you drop low from launch and climb back up it records much lower than you went, usually under the ground. This is the case for all the GPS devices I have tested and is clearly visible from just about all igc  files you might see from gliding flight.So GPS altitude while flying, unless you are cruising level with a motor, is not good for the pilot to refer to.If you recal  altimeter with gps at top of last climb you will arrive too low at goal every time.Changes in pressure are usually small, because we are not capable of flying in weather where we can fly across many isobars on the weather map. Very long distances are along the isobars anyway ;)So using a fixed barometric altitude from launch is usually best. Base map if it is flat, GPS averaged over time if it is sloping.On 2 Jul 2018 10:45 am, "Dr. Tilmann Bubeck" <notifications@github.com> wrote:You got it. :)

Two ideas:

it would be better to turn on SkyDrop some time before the start (e.g. when you walk to the starting point). This would give it a chance to recalibrate based on DEM whenever you cross a "flat" area. Apart from that, you are rightYes, it will recalibrate during flight with GPS 3d (every few hours). We can enter a comment into IGC to explain this. I think that a recalibration is better, than a wrong altitude... Maybe we find some better handling of glitches.

You should try it out and report, what you think should be done to improve.

**Comment by bubeck** [comment]
You can see VDOP, HDOP and PDOP in the GPS details view under "Menu".

**Comment by bluearcus** [comment]
I would like to be able to see them on a default startup screen which I look at as I'm getting ready...

Time
Battery %
GPS Satellites in view/GPS satellites used in fix
HDOP and VDOP
Altitude

Having to go into the menu to see GPS stats, not so good. If I don't like what I see on this screen, I can make my own calibration of my main baro altimeter (which should set the error to zero...)

**Comment by moyesboy** [comment]
It seems I can alter all the atimeters 1 thru 5 on screen, and possibly while flying too?
I would like to see one altimeter that is QNE, and not user settable. ie it displays the altitude being recorded in the igc file (1013.25mB = zero).
Having now fiddled with the altimeters I have now also run into the issue of not being able to set QNH1 to 1013.25 anymore... was there a solution to that? seems I cant get any altimeter to read exactly what is being recorded in the igc...
as per **[Fulminao](/Fulminao) ** commented [on 9 Feb 2016](#issuecomment-181832966)

**Comment by bubeck** [comment]
You should be able to use alt5 for exactly that if you haven't changed it in the setting. Try reset to factory.

**Comment by moyesboy** [comment]
I think it would be better to commit one altimeter to 1013.25mb, so it shows flight levels, and cannot be changed.I see that each one begins at 1013.25 but each one can be changed.... And the only way to get back to 1013.25 exactly is a factory reset...It coupld be named AltFL rather than Alti5 or whatever so its clear what it is and why it is not adjustable.On 12 Feb 2019 7:58 am, "Dr. Tilmann Bubeck" <notifications@github.com> wrote:You should be able to use alt5 for exactly that if you haven't changed it in the setting. Try reset to factory.

**Comment by ukeconsulting** [comment]
i use this feature because i like the "turn on and forget" use of the device but each time i use the web configurator i have to manually put the option back in the vario after update. 
is there a place where we can find the option in the web configurator ? i didn't find the checkbox in https://vps.skybean.eu/configurator/#/altimeters

**Comment by bubeck** [comment]
This could be a problem in the configurator. Which exact option should be saved? I don't understand this from your previous comment.

If I know exactly what is missing, we can fix it

**Comment by ukeconsulting** [comment]
The option to set the altitude 1 with GPS.

Cf 
https://image.noelshack.com/fichiers/2019/23/1/1559598573-img-20190603-233922.jpg
